src/closure.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright 2013-2017, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+// TODO: support marshaller.
+
+use std::mem;
+use std::ptr;
+use std::slice;
+
+use libc::{c_uint, c_void};
+
+use gobject_ffi;
+use translate::{ToGlibPtr, ToGlibPtrMut, Uninitialized, from_glib_none};
+use types::Type;
+use Value;
+use ToValue;
+
+glib_wrapper! {
+    pub struct Closure(Shared<gobject_ffi::GClosure>);
+
+    match fn {
+        ref => |ptr| {
+            gobject_ffi::g_closure_ref(ptr);
+            gobject_ffi::g_closure_sink(ptr);
+        },
+        unref => |ptr| gobject_ffi::g_closure_unref(ptr),
+    }
+}
+
+impl Closure {
+    pub fn new<F: Fn(&[Value]) -> Option<Value> + 'static>(callback: F) -> Self {
+        unsafe extern "C" fn marshal<F: Fn(&[Value]) -> Option<Value> + 'static>(_closure: *mut gobject_ffi::GClosure,
+            return_value: *mut gobject_ffi::GValue, n_param_values: c_uint, param_values: *mut gobject_ffi::GValue,
+            _invocation_hint: *mut c_void, marshal_data: *mut c_void)
+        {
+            let values = slice::from_raw_parts(param_values as *const _, n_param_values as usize);
+            let callback: Box<F> = Box::from_raw(marshal_data as *mut _);
+            let result = callback(values);
+            if !return_value.is_null() {
+                match result {
+                    Some(result) => *return_value = result.into_raw(),
+                    None => {
+                        let result = Value::uninitialized();
+                        *return_value = result.into_raw();
```

Comment:
There's no type it could be initialized with here. We just need to fill the return value with something here, just in case. And that something is a GValue initialized with all zeroes (i.e. G_TYPE_INVALID), something we can detect later and handle appropriately.